[15.0][FIX] purchase_discount_*: use hook method for supplierinfo

chienandalu opened the pull request:
From version 15 we've got this handy _prepare_supplier_info method so we don't have to do context injection juggling to add the discount info.

please review @victoralmau @sergio-teruel

pedrobaeza reviewed (changes requested):
It's better to not put 2 return in the method.

If there's no `vals["product_tmpl_id"]`, then you should check `vals["product_tmpl_id"]` and extract the template from it.

chienandalu commented:
> It's better to not put 2 return in the method.
> 
> If there's no `vals["product_tmpl_id"]`, then you should check `vals["product_tmpl_id"]` and extract the template from it.

The thing is supplierinfo records can be created either without any product info or without this context or both.

pedrobaeza commented:
Correct, but you still need to know the product template through the mechanism I'm talking about.

pedrobaeza commented:
@chienandalu I have rebased resolving conflicts (due to a new `.get("product_tmpl_id")` instead of `["product_tmpl_id"]`). Can you please check my comment to finish this?

chienandalu commented:
> If there's no vals["product_tmpl_id"], then you should check vals["product_tmpl_id"] and extract the template from it.

Do you mean from vals["product_id"]?

pedrobaeza commented:
Yes, sorry.

pedrobaeza commented:
@chienandalu can you please finish this?

chienandalu committed:
[FIX] purchase_discount: use hook method for supplierinfo

From version 15 we've got this handy _prepare_supplier_info method so we
don't have to do context injection juggling to add the discount info.

chienandalu commented:
The main issue was already fixed here: https://github.com/OCA/purchase-workflow/pull/2079

Anyway, I implemented your request, although I'm not completly sure it's needed for the case of this context (the creation of the supplierinfo from the method `_add_supplier_to_product`, which should be quite predictable

pedrobaeza commented on purchase_discount/models/product_supplierinfo.py:
```diff
@@ -27,9 +27,19 @@ def _get_po_to_supplierinfo_synced_fields(self):
-            product_tmpl_id = vals.get("product_tmpl_id")
-            po_line_map = self.env.context.get("po_line_map", {})
+        po_line_map = self.env.context.get("po_line_map")
+        if not po_line_map:
```
I don't understand what this exit condition does.

chienandalu replied:
Well, I was in my way to explain how this is a dict injected in context with the only purpose of passing the discount values on creation when I realized that in this version there's a hook method that makes this engineering unnecessary

pedrobaeza commented on purchase_discount/models/product_supplierinfo.py:
```diff
@@ -27,9 +27,19 @@ def _get_po_to_supplierinfo_synced_fields(self):
+        if not po_line_map:
+            return super().create(vals_list)
+        for vals in filter(
+            lambda x: x.get("product_tmpl_id", x.get("product_id")), vals_list
```
This filter seems now incorrect if we fallback to the other value.

chienandalu replied:
Not important now, but it just meant that we could go with that dictionary as we could get the template from the product_id

chienandalu committed:
[FIX] purchase_triple_discount: use hook method for supplierinfo

From version 15 we've got this handy _prepare_supplier_info method so we
don't have to do context injection juggling to add the discount info.

pedrobaeza reviewed (approved):
This is way better!

/ocabot merge patch

rousseldenis commented:
@chienandalu Do you plan to forwardport this in 16 ?

chienandalu commented:
The thing is that the methods changed completely and now there isn't an obvious hook, so maybe an extra PR to Odoo will be needed to try solve it...

rousseldenis commented:
> The thing is that the methods changed completely and now there isn't an obvious hook, so maybe an extra PR to Odoo will be needed to try solve it...

Ok. I'll maybe look at it

chienandalu commented:
Great! ping me if you want to when you're on it :smiley:
